healthcheck/handler.go

```diff
@@ -0,0 +1,109 @@
+package healthcheck
+
+import (
+	"encoding/json"
+	"github.com/ONSdigital/go-ns/log"
+	"github.com/pkg/errors"
+	"net/http"
+	"time"
+)
+
+const (
+	StatusOK       = "OK"
+	StatusCritical = "CRITICAL"
+	StatusWarning  = "WARNING"
+)
+
+// Handler returns the current health check of the app
+func (hc HealthCheck) Handler(w http.ResponseWriter, req *http.Request) {
+	now := time.Now().UTC()
+	ctx := req.Context()
+
+	var checks []Check
+
+	for _, client := range hc.Clients {
+		if client.Check != nil {
+			checks = append(checks, *client.Check)
+		}
+	}
+
+	hc.Status = hc.getStatus()
+	hc.Uptime = now.Sub(hc.StartTime)
+		hc.Checks =    checks
+
+
+	b, err := json.Marshal(hc)
+
+	if err != nil {
+		log.ErrorCtx(ctx, errors.Wrap(err, "failed to marshal json"), log.Data{"error": err, "health_check_response": hc})
+		return
+	}
+
+	_, err = w.Write(b)
+	if err != nil {
+		log.ErrorCtx(ctx, errors.Wrap(err, "failed to write bytes for http response"), nil)
+		return
+	}
+}
+
+// isAppStartingUp returns true or false based on if each client has had results back from at least a single check
+func (hc HealthCheck) isAppStartingUp() bool {
+	for _, client := range hc.Clients {
+		if client.Check == nil {
+			return true
+		}
+	}
+	return false
+}
+
+// getStatus returns a status as string as to the overall current apps health based on its dependent apps health
+func (hc HealthCheck) getStatus() string {
+	if hc.isAppStartingUp() {
+		return StatusWarning
+	}
+	return hc.isAppHealthy()
+}
+
+// isAppHealthy checks every clients check for their health then produces and returns a status for this apps health
+func (hc HealthCheck) isAppHealthy() string {
+	status := StatusOK
+	defer unlockAllClientChecks(hc.Clients)
+	for _, client := range hc.Clients {
+		client.mutex.Lock()
+		appHealth := hc.isCheckHealthy(client.Check)
+		if appHealth == StatusCritical {
```

Comment:
So initially I did exactly that ^^ but was pointed out to use defer instead in case something panics. However can't really use defers in loops - bad practice of so the interwebs says so instead created a new function to do them all in - but as pointed out here not great - so went back the drawing board and refactored this section of code to move everything from the loop into its own function.

Reply:
So I initially had it like that but it was pointed out that there could be a panic and so may not unlock. So I added a defer but instead however can get resource leaks if you have `defer`s in iterative loops. So I then created this function as an alternative - but I have also now refactored it further and separated out the functionality from this loop into its own function, in this function the loops are also handled, this should resolve that issue.